cpp/src/utilities/host_memory.cpp

```diff
@@ -62,69 +62,89 @@ class fixed_pinned_pool_memory_resource : public rmm::mr::device_memory_resource
-        // If the pool is exhausted, fall back to the upstream memory resource
+    try {
+      return pool_->allocate(stream, bytes);
+    } catch (...) {
+      CUDF_LOG_INFO("Pinned pool exhausted, falling back to new pinned allocation for %zu bytes",
+                    bytes);
+      // fall back to upstream
+      auto* ptr = upstream_mr_.allocate(stream, bytes);
+
+      {
+        std::unique_lock lock(fallback_allocations_mutex_);
+        fallback_allocations_.insert(ptr);
       }
-    }
 
-    return upstream_mr_.allocate(stream, bytes);
+      return ptr;
+    }
   }
 
   void do_deallocate(void* ptr, std::size_t bytes, rmm::cuda_stream_view stream) noexcept override
   {
-    if (bytes <= pool_size_ && ptr >= pool_begin_ && ptr < pool_end_) {
-      pool_->deallocate(stream, ptr, bytes);
-    } else {
-      upstream_mr_.deallocate(stream, ptr, bytes);
+    {
+      std::shared_lock lock(fallback_allocations_mutex_);
+      if (fallback_allocations_.find(ptr) == fallback_allocations_.end()) {
+        pool_->deallocate(stream, ptr, bytes);
+        return;
+      }
     }
+
+    {
+      std::unique_lock lock(fallback_allocations_mutex_);
+      fallback_allocations_.erase(ptr);
+    }
+    upstream_mr_.deallocate(stream, ptr, bytes);
   }
 
   [[nodiscard]] bool do_is_equal(device_memory_resource const& other) const noexcept override
   {
-    auto const* other_ptr = dynamic_cast<fixed_pinned_pool_memory_resource const*>(&other);
-    return other_ptr != nullptr && pool_ == other_ptr->pool_;
+    auto const* other_ptr = dynamic_cast<pinned_pool_with_fallback_memory_resource const*>(&other);
+    return other_ptr != nullptr && pool_ == other_ptr->pool_ && stream_ == other_ptr->stream_;
   }
 };
 
-static_assert(cuda::mr::resource_with<fixed_pinned_pool_memory_resource,
+static_assert(cuda::mr::resource_with<pinned_pool_with_fallback_memory_resource,
                                       cuda::mr::device_accessible,
                                       cuda::mr::host_accessible>,
               "Pinned pool mr must be accessible from both host and device");
 
 CUDF_EXPORT rmm::host_device_async_resource_ref& make_default_pinned_mr(
   std::optional<size_t> config_size)
 {
-  static fixed_pinned_pool_memory_resource mr = [config_size]() {
-    auto const size = [&config_size]() -> size_t {
+  static pinned_pool_with_fallback_memory_resource mr = [config_size]() {
+    auto const initial_size = [&config_size]() -> size_t {
       if (auto const env_val = getenv("LIBCUDF_PINNED_POOL_SIZE"); env_val != nullptr) {
         return std::atol(env_val);
       }
 
       if (config_size.has_value()) { return *config_size; }
 
       auto const total = rmm::available_device_memory().second;
-      // 0.5% of the total device memory, capped at 100MB
-      return std::min(total / 200, size_t{100} * 1024 * 1024);
+      // 0.5% of the total device memory, capped at 64MB
+      return std::min(total / 200, size_t{64} * 1024 * 1024);
```

Comment:
Wait why did we cap the size of the pinned pool at 64MB instead of 100MB? Because of the fallback?

Reply:
We decrease the starting size, but allow more than 10x of the original size before switching to new allocations.
If we make changes that increase pinned memory use, we can increase the initial size (and the max size will scale with it) for now I wanted a balance of low initial cost and increased max size.